### PR TITLE
Cover Seeing payment method instructions behat API scenario

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                     |
| License         | MIT                                                          |

### CODE REVIEW

File: features/checkout/paying_for_order/seeing_payment_method_instructions_on_thank_you_page.feature [hunk 1]
```diff
@@ -11,11 +11,19 @@ Feature: Seeing payment method instructions on thank you page
...
         Then I should see the thank you page
         And I should be informed with "Offline" payment method instructions
+
+    @api @no-ui
+    Scenario: Being informed about payment instructions
+        Given I am a logged in customer
+        And I have product "PHP T-Shirt" in the cart
+        When I proceed selecting "Offline" payment method
+        And I confirm my order
+        And I should be informed with "Offline" payment method instructions
```
Comment (jakubtobiasz) [inline features/checkout/paying_for_order/seeing_payment_method_instructions_on_thank_you_page.feature]: I'm wondering whether the information about the "thank you" page is important here. Maybe we can go with

```suggestion
    @ui @api
    Scenario: Being informed about payment instructions on thank you page
        Given I am a logged in customer
        And I have product "PHP T-Shirt" in the cart
        When I proceed selecting "Offline" payment method
        And I confirm my order
        Then I should be informed with "Offline" payment method instructions
```

wdyt?
cc: @GSadee @NoResponseMate

Comment (NoResponseMate) [reply features/checkout/paying_for_order/seeing_payment_method_instructions_on_thank_you_page.feature]: I'm fine with that, but swap the `on thank you page` from the scenario/feature/filename to `after checkout`

Comment (jakubtobiasz) [reply features/checkout/paying_for_order/seeing_payment_method_instructions_on_thank_you_page.feature]: Ah, yes, I missed that.

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 2]
```diff
@@ -553,6 +553,37 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
         );
     }
 
+    /**
+     * @Then I should be informed with :paymentMethod payment method instructions
+     */
+    public function iShouldBeInformedWithPaymentMethodInstructions(PaymentMethodInterface $paymentMethod)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]: Missing return type

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 3]
```diff
@@ -553,6 +553,37 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+                    $this->responseChecker->getValue(
+                        $paymentMethodResponse, 'instructions'),
+                        $paymentMethod->getInstructions(),
+                        sprintf('Payment method instructions should be equal to %s', $paymentMethod->getInstructions()
+                    )
+                );
+            }
+        }
+
+        Assert::true($paymentMethodFound, sprintf('Payment method %s not found in response.', $paymentMethod->getName()));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]:
```suggestion
        Assert::notEmpty($payments, 'No payments found in response.');

        $paymentMethodIri = $this->iriConverter->getIriFromItem($paymentMethod);
        foreach ($payments as $payment) {
            if ($payment['method'] === $paymentMethodIri) {
                $customRequest = $this->requestFactory->custom($payment['method'], HTTPRequest::METHOD_GET);
                $paymentMethodResponse = $this->client->executeCustomRequest($customRequest);

                Assert::same(
                    $this->responseChecker->getValue(
                        $paymentMethodResponse, 'instructions'),
                        $paymentMethod->getInstructions(),
                        sprintf('Payment method instructions should be equal to %s', $paymentMethod->getInstructions()
                    )
                );
                
                return;
            }
        }

        throw new \Exception(sprintf('Payment method %s not found in response.', $paymentMethod->getName()));
```
🤔

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 4]
```diff
@@ -553,6 +553,35 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+
+        Assert::notEmpty($payments, 'No payments found in response.');
+        $paymentMethodIri = $this->iriConverter->getIriFromItem($paymentMethod);
+        foreach ($payments as $payment) {
+            if ($payment['method'] === $paymentMethodIri) {
+                $customRequest = $this->requestFactory->custom($payment['method'], HTTPRequest::METHOD_GET);
+                $paymentMethodResponse = $this->client->executeCustomRequest($customRequest);
+                Assert::same(
+                    $this->responseChecker->getValue(
+                        $paymentMethodResponse, 'instructions'),
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]: Either
```suggestion
                    $this->responseChecker->getValue($paymentMethodResponse, 'instructions'),
```
or 
```suggestion
                    $this->responseChecker->getValue(
                        $paymentMethodResponse,
                        'instructions',
                    ),

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -553,6 +553,35 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+                    $paymentMethod->getInstructions(),
+                    sprintf('Payment method instructions should be equal to %s', $paymentMethod->getInstructions()
+                    )
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]:
```suggestion
                    sprintf('Payment method instructions should be equal to %s', $paymentMethod->getInstructions()),
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 6]
```diff
@@ -553,6 +553,35 @@ public function iShouldBeOnTheCheckoutCompleteStep(): void
...
+     */
+    public function iShouldBeInformedWithPaymentMethodInstructions(PaymentMethodInterface $paymentMethod): void
+    {
+        $response = $this->client->getLastResponse();
+        $payments = $this->responseChecker->getValue($response, 'payments');
+
+        Assert::notEmpty($payments, 'No payments found in response.');
+        $paymentMethodIri = $this->iriConverter->getIriFromItem($paymentMethod);
+        foreach ($payments as $payment) {
+            if ($payment['method'] === $paymentMethodIri) {
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]:
```suggestion
            if ($payment['method'] !== $paymentMethodIri) {
                continue;
            }
```
Would make it easier to read.

**Review by NoResponseMate — DISMISSED** [review]

File: features/checkout/paying_for_order/seeing_payment_method_instructions_after_checkout.feature [hunk 7]
```diff
@@ -11,11 +11,10 @@ Feature: Seeing payment method instructions on thank you page
...
-    @ui
-    Scenario: Being informed about payment instructions on thank you page
+    @ui @api
+    Scenario: Being informed about payment instructions
         Given I am a logged in customer
         And I have product "PHP T-Shirt" in the cart
         When I proceed selecting "Offline" payment method
         And I confirm my order
-        Then I should see the thank you page
         And I should be informed with "Offline" payment method instructions
```
Comment (GSadee) [inline features/checkout/paying_for_order/seeing_payment_method_instructions_after_checkout.feature:20]:
```suggestion
        Then I should be informed with "Offline" payment method instructions
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Kamil! :1st_place_medal:

**Pull request merged**